src/Confluent.Kafka/PartitionRequest.cs

```diff
@@ -0,0 +1,98 @@
+﻿//    Copyright 2019 Chad Jessup
+// 
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+// 
+//        http://www.apache.org/licenses/LICENSE-2.0
+// 
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+// 
+// 
+// Derived from: rdkafka-dotnet - C# Apache Kafka client library
+// 
+// Copyright (c) 2015-2016, Andreas Heider
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice,
+//    this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    ///     Request to retrieve the partition to send a message to.
+    /// </summary>
+    /// <remarks>
+    ///     A request for the partition is only generated if the <seealso cref="Partition"/> is set to <seealso cref="Partition.Any"/>.
+    /// </remarks>
+    /// <typeparam name="TKey">The type of the Key.</typeparam>
+    /// <typeparam name="TValue">The type of the Value.</typeparam>
+    public class PartitionRequest<TKey, TValue>
```

Comment:
I'm  not too fond of this name, the word "Request" is already used to indicate a protocol request, which this is not, and there's already existing nomenclature for this, which is Partitioner.

Reply:
I'm not too fond either...
How does ```SetPartionerHandler``` sound for the handler?  And I don't know if this class technically counts as the ```Partitioner``` itself. It's really just a class that holds the data to be passed to the handler...